contrib/go.mongodb.org/mongo-driver.v2/mongo: add support

jwilner opened the pull request:
### What does this PR do?

Adds support for go.mongodb.org/mongo-driver/v2. Exactly the same interface as
v1.

Because it's exactly the same interface, there are probably opportunities to factor 
out the code with v1 using e.g. type conversions. That said, that would limit the
ability of the code to evolve independently. Happy to factor it out if there's a 
preference.

### Motivation

No existing support.

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).
-->

- [ ] Changed code has unit tests for its functionality at or near 100% coverage.
- [ ] [System-Tests](https://github.com/DataDog/system-tests/) covering this feature have been added and enabled with the va.b.c-dev version tag.
- [ ] There is a benchmark for any new code, or changes to existing code.
- [ ] If this interacts with the agent in a new way, a system test has been added.
- [ ] Add an appropriate team label so this PR gets put in the right place for the release notes.
- [ ] Non-trivial go.mod changes, e.g. adding new modules, are reviewed by @DataDog/dd-trace-go-guild.
- [ ] For internal contributors, a matching PR should be created to the `v2-dev` branch and reviewed by @DataDog/apm-go.


Unsure? Have a question? Request a review!

darccio commented:
@jwilner As part of our upcoming release, we are reviewing all in-flight PRs. Please consider closing it or rebasing to `v2-dev` if you want to merge it for the next release after `v2.0.0`.

Also, thanks for contributing it. We'll review it soon.

jemiahw commented:
What's the status on this PR? Any idea when it might be reviewed and merged?

nikhilsinghcohesity commented:
Looking forward for this PR to be merged. Please let me know if i can also provide any kind of assistance with getting this merged. Thanks.

Teko012 commented:
@jwilner @darccio Any news on adding this?

darccio commented:
@Teko012 We should address the requested changes. We've been busy dogfooding and releasing `dd-trace-go` v2, but it's out now. We'll catch up with pending PRs like this one during the next weeks.

rarguelloF commented on contrib/go.mongodb.org/mongo-driver.v2/mongo/orchestrion.yml:
```diff
@@ -0,0 +1,21 @@
+# Unless explicitly stated otherwise all files in this repository are licensed
```
If we are adding orchestrion support, we should add an integration test as well (under `internal/orchestrion/_integration`).

rarguelloF commented on contrib/go.mongodb.org/mongo-driver.v2/mongo/orchestrion.yml:
```diff
@@ -0,0 +1,21 @@
+---
+# yaml-language-server: $schema=https://datadoghq.dev/orchestrion/schema.json
+meta:
+  name: gopkg.in/DataDog/dd-trace-go.v1/contrib/go.mongodb.org/mongo-driver.v2/mongo
```
```suggestion
  name: github.com/DataDog/dd-trace-go/contrib/go.mongodb.org/mongo-driver.v2/v2/mongo
```

jwilner replied:
Should be addressed.

rarguelloF commented on contrib/go.mongodb.org/mongo-driver.v2/mongo/option.go:
```diff
@@ -0,0 +1,67 @@
+	}
+}
+
+// WithAnalytics enables Trace Analytics for all started spans.
+func WithAnalytics(on bool) Option {
+	return func(cfg *config) {
+		if on {
+			cfg.analyticsRate = 1.0
+		} else {
+			cfg.analyticsRate = math.NaN()
+		}
+	}
+}
+
+// WithAnalyticsRate sets the sampling rate for Trace Analytics events
+// correlated to started spans.
+func WithAnalyticsRate(rate float64) Option {
+	return func(cfg *config) {
+		if rate >= 0.0 && rate <= 1.0 {
+			cfg.analyticsRate = rate
+		} else {
+			cfg.analyticsRate = math.NaN()
+		}
+	}
+}
```
```suggestion
```

This is a [legacy feature](https://docs.datadoghq.com/tracing/legacy_app_analytics/?tab=go) so there's no need to add these options for new integrations.

rarguelloF commented on contrib/go.mongodb.org/mongo-driver.v2/mongo/option.go:
```diff
@@ -0,0 +1,67 @@
+func defaults(cfg *config) {
+	cfg.serviceName = namingschema.ServiceNameOverrideV0(defaultServiceName, defaultServiceName)
+	cfg.spanName = namingschema.OpName(namingschema.MongoDBOutbound)
+	// cfg.analyticsRate = globalconfig.AnalyticsRate()
+	if internal.BoolEnv("DD_TRACE_MONGO_ANALYTICS_ENABLED", false) {
+		cfg.analyticsRate = 1.0
+	} else {
+		cfg.analyticsRate = math.NaN()
+	}
```
```suggestion
```

(same reason)

rarguelloF commented on contrib/go.mongodb.org/mongo-driver.v2/mongo/option.go:
```diff
@@ -0,0 +1,67 @@
+type config struct {
+	serviceName   string
+	spanName      string
+	analyticsRate float64
```
```suggestion
```

rarguelloF commented on contrib/go.mongodb.org/mongo-driver.v2/mongo/option.go:
```diff
@@ -0,0 +1,67 @@
+type Option func(*config)
+
+func defaults(cfg *config) {
+	cfg.serviceName = namingschema.ServiceNameOverrideV0(defaultServiceName, defaultServiceName)
+	cfg.spanName = namingschema.OpName(namingschema.MongoDBOutbound)
```
this package `namingschema` does not exist anymore, you can check in the existing mongodb integration how this is done in v2.

rarguelloF commented on contrib/go.mongodb.org/mongo-driver.v2/mongo/mongo_test.go:
```diff
@@ -0,0 +1,167 @@
+	})
+}
+
+func TestNamingSchema(t *testing.T) {
+	genSpans := namingschematest.GenSpansFn(func(t *testing.T, serviceOverride string) []mocktracer.Span {
+		var opts []Option
+		if serviceOverride != "" {
+			opts = append(opts, WithServiceName(serviceOverride))
+		}
+		mt := mocktracer.Start()
+		defer mt.Stop()
+
+		addr := fmt.Sprintf("mongodb://localhost:27017/?connect=direct")
+		mongopts := options.Client()
+		mongopts.Monitor = NewMonitor(opts...)
+		mongopts.ApplyURI(addr)
+		client, err := mongo.Connect(context.Background(), mongopts)
+		require.NoError(t, err)
+		_, err = client.
+			Database("test-database").
+			Collection("test-collection").
+			InsertOne(context.Background(), bson.D{{Key: "test-item", Value: "test-value"}})
+		require.NoError(t, err)
+
+		return mt.FinishedSpans()
+	})
+	namingschematest.NewMongoDBTest(genSpans, "mongo")(t)
+}
```
```suggestion
```

These tests have been moved to `instrumentation/internal/namingschematest`

jwilner commented on contrib/go.mongodb.org/mongo-driver.v2/mongo/mongo.go:
```diff
@@ -0,0 +1,132 @@
+func init() {
+	telemetry.LoadIntegration(componentName)
+	tracer.MarkIntegrationImported("go.mongodb.org/mongo-driver/v2")
+}
```
Should be addressed by update.

jwilner commented on contrib/go.mongodb.org/mongo-driver.v2/mongo/option.go:
```diff
@@ -0,0 +1,67 @@
+		cfg.analyticsRate = 1.0
+	} else {
+		cfg.analyticsRate = math.NaN()
+	}
```
Addressed

jwilner commented:
@darccio @rarguelloF I recreate the PR against the new v2 code, removed the analytics rate cfg, and added an orchestrion integration test. I think recreating the PR against the new v2 code addressed most of your comments.
